File: scripts/sandbox-seed/_finance-match.js
// Match /tmp/finance-rows.json companies against sandbox accounts.
// Normalized name comparison. Reports matched/unmatched. No writes.

require('dotenv').config({ path: '.env.sandbox' });
const { Client } = require('pg');
const fs = require('fs');

const norm = (s) => (s || '').toLowerCase().replace(/l\.l\.c\.?/g, 'llc').replace(/[^a-z0-9]/g, '');

(async () => {
  const rows = JSON.parse(fs.readFileSync('/tmp/finance-rows.json', 'utf8'));
  const c = new Client({ connectionString: process.env.SUPABASE_DB_URL, ssl: { rejectUnauthorized: false } });
  await c.connect();

  const acc = await c.query(`
    SELECT id, company_name, ein_number, status, account_type
    FROM accounts
    WHERE account_type IN ('Client','Managed','One-Time') OR account_type IS NULL
  `);

  const byNormName = new Map();
  for (const a of acc.rows) {
    byNormName.set(norm(a.company_name), a);
  }

  const matched = [];
  const unmatched = [];
  for (const r of rows) {
    const key = norm(r.company);
    const hit = byNormName.get(key);
    if (hit) {
      matched.push({
        company: r.company,
        amount: r.amount,
        paid_status: r.paid_status,
        sandbox_account_id: hit.id,
        sandbox_name: hit.company_name,
        sandbox_status: hit.status,
        sandbox_account_type: hit.account_type,
      });
    } else {
      unmatched.push({ company: r.company, amount: r.amount, paid_status: r.paid_status });
    }
  }

  fs.writeFileSync('/tmp/finance-match.json', JSON.stringify({ matched, unmatched }, null, 2));
  console.log(`matched:   ${matched.length}`);
  console.log(`unmatched: ${unmatched.length}`);
  if (unmatched.length) {
    console.log('\n=== UNMATCHED ===');
    for (const r of unmatched) console.log(`  ${r.company!=null?r.company:'(null)'}   amount=${r.amount}   status=${r.paid_status}`);
  }
  // matched distribution by sandbox_account_type + sandbox_status
  const dist = {};
  for (const m of matched) {
    const k = `${m.sandbox_account_type}|${m.sandbox_status}`;
    dist[k] = (dist[k] || 0) + 1;
  }
  console.log('\n=== matched distribution (sandbox account_type | status) ===');
  for (const [k,v] of Object.entries(dist).sort((a,b)=>b[1]-a[1])) console.log(`  ${v}  ${k}`);

  await c.end();
})().catch(e => { console.error('FAIL:', e.message); process.exit(1); });

File: scripts/sandbox-seed/_finance-match.py
"""
Match the 170 'paid/overdue' rows from the Managed Clients finance xlsx
against sandbox accounts by normalized company name.

Output: /tmp/finance-match.json with
  matched:   [ {company, amount, paid_status, sandbox_account_id, sandbox_name} ]
  unmatched: [ {company, amount, paid_status, reason} ]

No DB writes. Just a diff.
"""
import json, os, re, sys
import pandas as pd

def norm(s):
    if s is None: return ''
    s = str(s).lower()
    s = s.replace('l.l.c.', 'llc').replace(' l.l.c', ' llc')
    s = re.sub(r'[^a-z0-9]', '', s)
    return s

# --- Read xlsx and filter to paid/overdue rows ---
df = pd.read_excel('/Users/10225office/Downloads/Managed Clients-Table 1.xlsx', sheet_name=0, header=None)
data = df.iloc[1:].reset_index(drop=True)
# drop the header row that sits inside the data
data = data[data[4] != 'Acct Status'].reset_index(drop=True)

# col 1 = company name, col 5 = status, col 6 = amount
rows = []
for _, r in data.iterrows():
    name = r[1]
    status_raw = r[5]
    amount_raw = r[6]
    status = str(status_raw).strip().lower() if pd.notna(status_raw) else ''
    is_paid = status in ('paid', 'overdue, paid')
    is_overdue = status == 'overdue'
    if not (is_paid or is_overdue):
        continue
    amount = None
    if pd.notna(amount_raw):
        try:
            amount = float(amount_raw)
        except Exception:
            amount = None
    rows.append({
        'company': str(name).strip(),
        'amount': amount,
        'paid_status': 'paid' if is_paid else 'overdue',
    })

print(f'xlsx paid+overdue rows: {len(rows)}')
paid_ct = sum(1 for r in rows if r['paid_status'] == 'paid')
print(f'  paid:    {paid_ct}')
print(f'  overdue: {len(rows) - paid_ct}')

# --- Query sandbox ---
# We don't have sandbox access from python; dump the rows and let node do the match.
with open('/tmp/finance-rows.json', 'w') as f:
    json.dump(rows, f, indent=2)
print('rows -> /tmp/finance-rows.json')
